File: bindings/python/examples/how_tos/account_output/state_transition.py
import os

from dotenv import load_dotenv

from iota_sdk import Wallet, FilterOptions, utf8_to_hex

load_dotenv()

# In this example we will update the state metadata of an account output.

NEW_STATE_METADATA = 'updated state metadata 1'

if 'WALLET_DB_PATH' not in os.environ:
    raise Exception(".env WALLET_DB_PATH is undefined, see .env.example")

if 'STRONGHOLD_PASSWORD' not in os.environ:
    raise Exception(".env STRONGHOLD_PASSWORD is undefined, see .env.example")

if 'EXPLORER_URL' not in os.environ:
    raise Exception(".env EXPLORER_URL is undefined, see .env.example")

wallet = Wallet(os.environ['WALLET_DB_PATH'])

account = wallet.get_account('Alice')

# Sync account with the node
balance = account.sync()

if len(balance.accounts) == 0:
    raise Exception("No Account available in account 'Alice'")

account_id = balance.accounts[0]

account_output_data = account.unspent_outputs(
    FilterOptions(accountIds=[account_id]))[0]
print(
    f"Account {account_id} found in unspent output: {account_output_data.outputId}")

account_output = account_output_data.output
updated_account_output = wallet.get_client().build_account_output(
    account_id,
    unlock_conditions=account_output.unlockConditions,
    state_index=int(account_output.stateIndex) + 1,
    state_metadata=utf8_to_hex(NEW_STATE_METADATA),
    foundry_counter=account_output.foundryCounter,
    immutable_features=account_output.immutableFeatures,
    features=account_output.features,
)

wallet.set_stronghold_password(os.environ["STRONGHOLD_PASSWORD"])

print('Sending transaction...')
transaction = account.send_outputs([updated_account_output])
print(f'Transaction sent: {transaction.transactionId}')

# Wait for transaction to get included
blockId = account.retry_transaction_until_included(
    transaction.transactionId)
print(f'Block included: {os.environ["EXPLORER_URL"]}/block/{blockId}')

File: bindings/nodejs/examples/how_tos/account_output/state-transition.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { AccountOutput, Wallet, initLogger, utf8ToHex } from '@iota/sdk';

// This example uses secrets in environment variables for simplicity which should not be done in production.
require('dotenv').config({ path: '.env' });

// Run with command:
// yarn run-example ./how_tos/account_output/state-transition.ts

const NEW_STATE_METADATA = 'updated state metadata 1';

// In this example we will update the state metadata of an account output.
async function run() {
    initLogger();
    if (!process.env.FAUCET_URL) {
        throw new Error('.env FAUCET_URL is undefined, see .env.example');
    }
    if (!process.env.WALLET_DB_PATH) {
        throw new Error('.env WALLET_DB_PATH is undefined, see .env.example');
    }
    if (!process.env.STRONGHOLD_PASSWORD) {
        throw new Error(
            '.env STRONGHOLD_PASSWORD is undefined, see .env.example',
        );
    }
    try {
        // Create the wallet
        const wallet = new Wallet({
            storagePath: process.env.WALLET_DB_PATH,
        });

        // Get the account we generated with `01-create-wallet`
        const account = await wallet.getAccount('Alice');

        // May want to ensure the account is synced before sending a transaction.
        const balance = await account.sync();

        if (balance.accounts.length == 0) {
            throw new Error(`No Alias available in account 'Alice'`);
        }

        // We try to update the first account output in the account
        const accountId = balance.accounts[0];

        const accountOutputData = (
            await account.unspentOutputs({ accountIds: [accountId] })
        )[0];
        console.log(
            `Alias ${accountId} found in unspent output: ${accountOutputData.outputId}`,
        );

        const accountOutput = accountOutputData.output as AccountOutput;

        const updatedAccountOutput = await (
            await wallet.getClient()
        ).buildAccountOutput({
            accountId,
            unlockConditions: accountOutput.unlockConditions,
            stateIndex: accountOutput.stateIndex + 1,
            stateMetadata: utf8ToHex(NEW_STATE_METADATA),
            foundryCounter: accountOutput.foundryCounter,
            immutableFeatures: accountOutput.immutableFeatures,
            features: accountOutput.features,
        });

        await wallet.setStrongholdPassword(process.env.STRONGHOLD_PASSWORD);

        console.log('Sending transaction...');

        const transaction = await account.sendOutputs([updatedAccountOutput]);
        console.log(`Transaction sent: ${transaction.transactionId}`);

        // Wait for transaction to get included
        const blockId = await account.reissueTransactionUntilIncluded(
            transaction.transactionId,
        );
        console.log(
            `Block included: ${process.env.EXPLORER_URL}/block/${blockId}`,
        );
    } catch (error) {
        console.log('Error: ', error);
    }
    process.exit(0);
}

run();
